### DISCUSSION TITLE
Split DTOs on ServiceStack client

### DISCUSSION THREAD

**Discussion by PedroJBrito** [ServiceStack/Discuss#12] [Q&A]
Hi guys,

my question is very straightforward. Is there any way of spliting the generated DTO files on client side in multiple DTOs, like one for each service model?

Thank you in advance for your answer.

**Comment by mythz** [comment]
Not from a single command, but you can create multiple .dtos files each with their own custom IncludeTypes which specifies which types you want in each file:

https://docs.servicestack.net/csharp-add-servicestack-reference.html#includetypes

Then you can update them all with a single command, e.g:

    $ x csharp

Although if it's just C# DTOs you want to do this for, you're better off structuring the ServiceModel in the layout you want then just xcopying the *.cs files to your client project, but typically it'd be easier to just reference the ServiceModel project directly in your client App.

**Reply by PedroJBrito** [reply]
Hi mythz,

first of all thank you for your prompt answer. What you suggest is something that I've already considered, I was wondering if it would be a faster approach. One more question please, is any way to create a dto file for each endpoint (Route classes) with all its related classes without having to type each of them in the includeTypes attribute?

Thanks agains.

**Reply by mythz** [reply]
There isn't and it's not a use-case we'll ever optimize for considering the purpose of client DTOs is to maintain them in sync with the server, i.e. to seamlessly update them whenever they're changed. You shouldn't be modifying local DTOs on the client, they default to partial classes so any enhancements/customizations should only be done in custom partial classes independent from the generated DTOs so changes are preserved across DTO updates. So I'm not clear on the motivation for wanting to maintain the generated DTOs themselves in different classes and not just copy them over from your ServiceModel project where they're already maintained in your preferred structure, or just reference the ServiceModel project directly?

You can download all the DTOs for a single API, e.g. by clicking on the [Code tab in API Explorer](https://docs.servicestack.net/api-explorer#code-tab), you wouldn't be able to automate this for each API because any shared types used in multiple APIs would be redefined which would cause build conflicts, which is what you would need to use IncludeTypes for, so you can precisely define which DTOs should be maintained in which .dtos file.

**Reply by PedroJBrito** [reply]
Hello again,

the problem we're facing and that motivate this discussion about split the dto file is that the project is quite big and was developed over several years, and now we're making a client redesign with different technologies and we took servicestack client to communicate with the old developed API. However the generated dto has some duplicated identifiers with general names like OrderBy or FilterBy, which are classes with the same name but different value, and a solution we thought to solve this problem was to split the dtos. Do you have any suggestion on how to solve this problem of duplicate names?

Thank you again.

**Reply by mythz** [reply]
If this is C# the best solution is to reference the ServiceModel project or ServiceModel.dll binary in order to be able to use non unique named DTOs.

You're going to have conflicts with other languages that can't utilize .NET namespaces where DTOs would need to be uniquely named. Alternative ways to generate different DTOs based on groups is to [Include All Types within a C# namespace](https://docs.servicestack.net/csharp-add-servicestack-reference#includetypes), e.g:

```csharp
/* Options:
IncludeTypes: MyApp.ServiceModel.Admin/*
*/
```

Or [Include All Services in a Tag Group](https://docs.servicestack.net/csharp-add-servicestack-reference#include-all-services-in-a-tag-group) by using the [[Tag] attribute](https://docs.servicestack.net/api-design#group-services-by-tag), e.g:

```csharp
/* Options:
IncludeTypes: {web},{mobile}
*/
```

Which will work if you don't need to use multiple conflict named types within the same client, otherwise you'll need to rename the DTOs so they're all unique.
